### PR TITLE
734 large dice

### LINKED ISSUE #734: Fix Dice Score
## Subject of the issue

@YipengHu obeserved dice score > 1, which is clearly a bug.

It might be caused by https://github.com/DeepRegNet/DeepReg/commit/3a457eef313f4bee61d16e899d15a87e6c831e91.
where we somehow changed reduce_sum to mean, but as we are adding EPS to the numerator/denominator, the reduce mean may lead to values << 1 and therefore the result is no longer accurate.

### PR DESCRIPTION
**Pull request by mathpluscode**
# Description

The main issue that dice score > 1 because the added metrics were not normalized by the global batch size.

In this PR, we also double-checked/refactored the implementation of several loss functions and added extensive tests to verify.

Fixes #734

## Type of change

What types of changes does your code introduce to DeepReg?

_Please check the boxes that apply after submitting the pull request._

- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] Code style update (formatting, renaming)
- [x] Refactoring (no functional changes, no api changes)
- [ ] Documentation Update (fix or improvement on the documentation)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Other (if none of the other choices apply)

## Checklist

_Please check the boxes that apply after submitting the pull request._

_If you're unsure about any of them, don't hesitate to ask. We're here to help! This is
simply a reminder of what we are going to look for before merging your code._

- [x] I have
      [installed pre-commit](https://deepreg.readthedocs.io/en/latest/contributing/setup.html)
      using `pre-commit install` and formatted all changed files. If you are not
      certain, run `pre-commit run --all-files`.
- [x] My commits' message styles matches
      [our requested structure](https://deepreg.readthedocs.io/en/latest/contributing/commit.html),
      e.g. `Issue #<issue number>: detailed message`.
- [x] I have updated the
      [change log file](https://github.com/DeepRegNet/DeepReg/blob/main/CHANGELOG.md)
      regarding my changes.
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)

### CODE REVIEW

**Comment by mathpluscode** [comment]
@YipengHu feel free to try again?

**Comment by mathpluscode** [comment]
@YipengHu I checked the implementation of `loss` and `metric`, i.e.

- MSE loss https://github.com/tensorflow/tensorflow/blob/v2.4.1/tensorflow/python/keras/losses.py#L267-L322
- MSE metric https://github.com/tensorflow/tensorflow/blob/v2.4.1/tensorflow/python/keras/metrics.py#L2378-L2410

They are different classes calling the same functions.

Then regarding the value under multiple GPUs, the doc is here https://www.tensorflow.org/tutorials/distribute/custom_training#define_the_loss_function

So, in short, there are a lot of work to be done to make it right lol

**Comment by mathpluscode** [comment]
For now, I removed binary dice from metrics, as we were using `loss` class as `metric`.

Will transform existing loss functions and add metric back.

File: deepreg/dataset/loader/interface.py [hunk 1]
```diff
@@ -91,12 +91,12 @@ def get_dataset_and_preprocess(
         data_augmentation: Optional[Union[List, Dict]] = None,
     ) -> tf.data.Dataset:
         """
-        :param training: bool, indicating if it's training or not
-        :param batch_size: int, size of mini batch
-        :param repeat: bool, indicating if we need to repeat the dataset
-        :param shuffle_buffer_num_batch: int, when shuffling,
+        :param training: indicating if it's training or not
+        :param batch_size: total number of samples consumed per step.
```
Comment (YipengHu) [inline deepreg/dataset/loader/interface.py:95]: i'd be explicit to add "... per step, over all GPUs"

File: deepreg/loss/image.py [hunk 2]
```diff
@@ -63,7 +64,7 @@ def __init__(
         self,
         num_bins: int = 23,
         sigma_ratio: float = 0.5,
-        reduction: str = tf.keras.losses.Reduction.SUM,
+        reduction: str = tf.keras.losses.Reduction.NONE,
```
Comment (YipengHu) [inline deepreg/loss/image.py:67]: will test this with experiment in https://github.com/YipengHu/DeepReg/tree/dev/experiments/mrus2

**Review by YipengHu — APPROVED** [review]

File: deepreg/dataset/loader/interface.py (hunk 1 above)
Comment (mathpluscode) [reply deepreg/dataset/loader/interface.py]: added

**Review by kate-sann5100 — APPROVED** [review]
